File: 2023/day15/java/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {

    record Lens(String label, int focalLength) {}

    public static void main(String[] args) throws IOException {
        String input = Files.readString(Path.of("../input.txt"))
                .strip()
                .replace("\n", "");
        String[] steps = input.split(",");

        System.out.println("Part 1: " + part1(steps));
        System.out.println("Part 2: " + part2(steps));
    }

    static int hash(String s) {
        int current = 0;
        for (char c : s.toCharArray()) {
            current = ((current + c) * 17) % 256;
        }
        return current;
    }

    static int part1(String[] steps) {
        return Arrays.stream(steps)
                .mapToInt(Solution::hash)
                .sum();
    }

    static int part2(String[] steps) {
        @SuppressWarnings("unchecked")
        List<Lens>[] boxes = new ArrayList[256];
        for (int i = 0; i < 256; i++) {
            boxes[i] = new ArrayList<>();
        }

        for (String step : steps) {
            if (step.contains("=")) {
                String[] parts = step.split("=");
                String label = parts[0];
                int focalLength = Integer.parseInt(parts[1]);
                int boxNum = hash(label);

                // Find and replace, or add
                boolean found = false;
                for (int i = 0; i < boxes[boxNum].size(); i++) {
                    if (boxes[boxNum].get(i).label().equals(label)) {
                        boxes[boxNum].set(i, new Lens(label, focalLength));
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    boxes[boxNum].add(new Lens(label, focalLength));
                }
            } else {
                // Remove operation (ends with '-')
                String label = step.substring(0, step.length() - 1);
                int boxNum = hash(label);
                boxes[boxNum].removeIf(lens -> lens.label().equals(label));
            }
        }

        // Calculate focusing power
        int total = 0;
        for (int boxNum = 0; boxNum < 256; boxNum++) {
            for (int slot = 0; slot < boxes[boxNum].size(); slot++) {
                Lens lens = boxes[boxNum].get(slot);
                total += (boxNum + 1) * (slot + 1) * lens.focalLength();
            }
        }
        return total;
    }
}

File: 2023/day15/rust/solution.rs
use std::fs;

/// Compute the HASH algorithm on a string
fn hash(s: &str) -> usize {
    s.bytes().fold(0, |acc, b| ((acc + b as usize) * 17) % 256)
}

/// Parse a step into (label, operation, focal_length)
/// operation is '-' for remove, '=' for add/replace
fn parse_step(step: &str) -> (&str, char, Option<u8>) {
    if let Some(label) = step.strip_suffix('-') {
        (label, '-', None)
    } else {
        let parts: Vec<&str> = step.split('=').collect();
        let label = parts[0];
        let focal_length = parts[1].parse::<u8>().unwrap();
        (label, '=', Some(focal_length))
    }
}

/// Part 1: Sum of HASH values for all steps
fn part1(input: &str) -> usize {
    input
        .trim()
        .split(',')
        .map(|step| hash(step))
        .sum()
}

/// Part 2: HASHMAP procedure and focusing power calculation
fn part2(input: &str) -> usize {
    // 256 boxes, each containing an ordered list of (label, focal_length)
    let mut boxes: Vec<Vec<(&str, u8)>> = vec![Vec::new(); 256];

    for step in input.trim().split(',') {
        let (label, op, focal_length) = parse_step(step);
        let box_num = hash(label);

        match op {
            '-' => {
                // Remove lens with matching label
                boxes[box_num].retain(|(l, _)| *l != label);
            }
            '=' => {
                let fl = focal_length.unwrap();
                // Check if lens with same label exists
                if let Some(pos) = boxes[box_num].iter().position(|(l, _)| *l == label) {
                    // Replace existing lens
                    boxes[box_num][pos] = (label, fl);
                } else {
                    // Add new lens at end
                    boxes[box_num].push((label, fl));
                }
            }
            _ => unreachable!(),
        }
    }

    // Calculate focusing power
    boxes
        .iter()
        .enumerate()
        .map(|(box_num, lenses)| {
            lenses
                .iter()
                .enumerate()
                .map(|(slot, (_, focal_length))| {
                    (box_num + 1) * (slot + 1) * (*focal_length as usize)
                })
                .sum::<usize>()
        })
        .sum()
}

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");

    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
